docs/DifferentiableProgramming.md

```diff
@@ -1477,10 +1479,111 @@ provided with a transpose (e.g. scalar multiplication, matrix transposition, and
 matrix multiplication), because gradients can only be computed when the
 differential can be transposed.
 
-To make a function a linear map, use a `@transpose` attribute on the transpose
-function. If the function is only linear with respect to certain parameters, use
-a wrt: clause to indicate parameter positions, e.g. `@transpose(of: ..., wrt:
-(self, 0))`.
+To make an original function be linear, define a transpose function with a
+`@transpose` attribute that specifies the original function.
+
+##### Typing rules
+
+A function declaration does not have a fixed transpose type. This is because
+there can be multiple transpose functions that transpose the original function
+differently, e.g. with respect to different parameters, transposing under
+different generic constraints, etc.
+
+Given an original function declaration, a transpose function's type is
+determined from the following configurations:
+- Parameters to transpose with respect to.
+- Additional generic constraints that make the original function linear.
+
+The type of the transpose function under such configurations is a
+function that takes one argument whose type is the original function's result
+type and returns results that correspond to each original function parameter
+that is transposed with respect to. This definition, however, is a rough
+definition because there are differences among top-level functions, instance
+methods, and static methods.
+
+###### Linearity parameters
+
+Linearity parameters are parameters with respect to which a function is linear.
+The `@transpose` attribute accepts a `wrt:` argument which specifies a set of
+linearity parameters of the original function. If `wrt:` is not specified,
+linearity parameters default to all parameters. A `wrt:` argument in
+`@derivative` attributes can be a parameter index or a tuple of multiple
+parameter indices. All linearity parameters must have a type that conforms to
+both `Differentiable` and `AdditiveArithmetic`.
```

Comment:
The linearity requirement should be `T: Differentiable, T == T.TangentVector`.
```suggestion
`Differentiable` and satisfies `Self == TangentVector`.
```

Implementation fixed in https://github.com/apple/swift/pull/28927.

Reply:
Not having `Self == TangentVector` was by design, as the [`@differentiable` function type attribute](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#the-differentiable-function-type-attribute) and the [implied generic constraints](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#implied-generic-constraints) sections don't have this constraint. I think however adding `T.TangentVector == T` is the right thing to do, so I fixed them.